src/main/java/gregtech/common/metatileentities/storage/MetaTileEntityWorkbench.java

```diff
@@ -198,34 +196,246 @@ public void clearMachineInventory(@NotNull List<@NotNull ItemStack> itemBuffer)
+        return SlotGroupWidget.builder()
+                .matrix("XXX",
+                        "XXX",
+                        "XXX")
+                .key('X', i -> CraftingInputSlot.create(this.recipeLogic, this.craftingGrid, i)
+                        .changeListener((newItem, onlyAmountChanged, client, init) -> {
+                            if (!init) {
+                                this.recipeLogic.updateCurrentRecipe();
+                            }
+                        })
+                        .background(GTGuiTextures.SLOT))
+                .build()
+                .child(new ButtonWidget<>()
+                        .margin(2)
+                        .size(8)
+                        .topRel(0f)
+                        .rightRel(0f, 0, 1f)
+                        .background(GTGuiTextures.BUTTON_CLEAR_GRID)
+                        .addTooltipLine(IKey.lang("gregtech.machine.workbench.clear_grid"))
+                        .disableHoverBackground()
+                        .onMousePressed(mouseButton -> {
+                            this.recipeLogic.clearCraftingGrid();
+                            return true;
+                        }));
+    }
+
+    public IWidget createCraftingOutput(PosGuiData guiData, PanelSyncManager syncManager) {
+        var amountCrafted = new IntSyncValue(this::getItemsCrafted, this::setItemsCrafted);
+        syncManager.syncValue("amount_crafted", amountCrafted);
+        amountCrafted.updateCacheFromSource(true); // todo remove on mui2 rc3
+
+        return Flow.column()
+                .size(54)
+                .child(new CraftingOutputSlot(amountCrafted, this)
+                        .marginTop(18)
+                        .background(GTGuiTextures.SLOT.asIcon().size(22))
+                        .marginBottom(4))
+                .child(IKey.dynamic(() -> TextFormattingUtil.formatLongToCompactString(amountCrafted.getIntValue(), 5))
+                        .alignment(Alignment.Center)
+                        .asWidget().widthRel(1f));
+    }
+
+    public IWidget createRecipeMemoryGrid(PanelSyncManager syncManager) {
+        return SlotGroupWidget.builder()
+                .matrix("XXX",
+                        "XXX",
+                        "XXX")
+                .key('X', i -> new RecipeMemorySlot(this.recipeMemory, i)
+                        .background(GTGuiTextures.SLOT))
+                .build().right(0);
+    }
+
+    public IWidget createInventoryPage(PanelSyncManager syncManager) {
+        if (this.connectedInventory.getSlots() == 0) {
+            return Flow.column()
+                    .debugName("inventory page - empty")
+                    .leftRel(0.5f)
+                    .padding(2)
+                    .height(18 * 6)
+                    .width(18 * 8 + 4)
+                    .background(GTGuiTextures.DISPLAY);
+        }
+
+        // this is actually supposed to include the tool and storage inventory
+        // but that causes problems
```

Comment:
eh, I don't think that the inventory page is supposed to contain the tool and internal inventory. It is really just a connected inventory page. So this comment is kinda not needed.

Reply:
the older inventory page used to include the tool and internal inventory as well as the connected inventory.